### PR TITLE
L4-data-structures

### PR DESCRIPTION
**Pull request by Angelbt12**
finished exercises

### CODE REVIEW

File: src/ab/js/l4/A-ist.js [hunk 1]
```diff
@@ -0,0 +1,32 @@
+function arrayToList(array) {
+  let list = null;
+  for (let i = array.length - 1; i >= 0; i--) {
```
Comment (carin91) [inline src/ab/js/l4/A-ist.js:3]: refactor using i++

File: src/ab/js/l4/A-ist.js [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+    list = { value: array[i], rest: list };
+  }
+  return list;
+}
+
+function listToArray(list) {
+  let array = [];
+  for (let node = list; node; node = node.rest) {
```
Comment (carin91) [inline src/ab/js/l4/A-ist.js:11]: avoid the coercion

File: src/ab/js/l4/A-ist.js [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+}
+
+function prepend(element, list) {
+  return { value: element, rest: list };
+}
+
+function nth(list, index) {
+  if (!list) return undefined;
+  else if (index === 0) return list.value;
+  else return nth(list.rest, index - 1);
```
Comment (carin91) [inline src/ab/js/l4/A-ist.js:24]: refactor to a non-recursive function

File: src/ab/js/l4/Deep-comparison.js [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+  ) {
+    let keysA = Object.keys(a);
+    let keysB = Object.keys(b);
+
+    if (keysA.length !== keysB.length) {
+      return false;
+    }
+
+    for (let key of keysA) {
+      if (!keysB.includes(key) || !deepEqual(a[key], b[key])) {
```
Comment (carin91) [inline src/ab/js/l4/Deep-comparison.js:20]: refactor with the and operator

File: src/ab/js/l4/Reversing-an-array.js [hunk 5]
```diff
@@ -0,0 +1,21 @@
+function reverseArray(array) {
+  let newArray = [];
+  for (let i = array.length - 1; i >= 0; i--) {
```
Comment (carin91) [inline src/ab/js/l4/Reversing-an-array.js:3]: refactor with i++

File: src/ab/js/l4/Reversing-an-array.js [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+    newArray.push(array[i]);
+  }
+  return newArray;
+}
+
+function reverseArrayInPlace(array) {
+  for (let i = 0; i < Math.floor(array.length / 2); i++) {
```
Comment (carin91) [inline src/ab/js/l4/Reversing-an-array.js:10]: refactor using i--

File: src/ab/js/l4/The-sum-of-a-range.js [hunk 7]
```diff
@@ -0,0 +1,21 @@
+function range(start, end, step = 1) {
+  let array = [];
+  if (step > 0) {
+    for (let i = start; i <= end; i += step) {
+      array.push(i);
+    }
+  } else {
```
Comment (carin91) [inline src/ab/js/l4/The-sum-of-a-range.js:7]: refactor to avoid the else statement

**Review by carin91 — CHANGES_REQUESTED** [review]

File: src/ab/js/l4/A-ist.js (hunk 1 above)
Comment (Angelbt12) [reply src/ab/js/l4/A-ist.js]: ok

File: src/ab/js/l4/A-ist.js (hunk 2 above)
Comment (Angelbt12) [reply src/ab/js/l4/A-ist.js:11]: ok

File: src/ab/js/l4/A-ist.js (hunk 3 above)
Comment (Angelbt12) [reply src/ab/js/l4/A-ist.js:24]: ok

File: src/ab/js/l4/Deep-comparison.js (hunk 4 above)
Comment (Angelbt12) [reply src/ab/js/l4/Deep-comparison.js:20]: ok

File: src/ab/js/l4/Reversing-an-array.js (hunk 5 above)
Comment (Angelbt12) [reply src/ab/js/l4/Reversing-an-array.js:3]: ok

File: src/ab/js/l4/The-sum-of-a-range.js (hunk 7 above)
Comment (Angelbt12) [reply src/ab/js/l4/The-sum-of-a-range.js:7]: ok

**Review by carin91 — APPROVED** [review]

**Pull request merged**